IO.cpp

```diff
@@ -0,0 +1,225 @@
+/*
+    microsoft-oms-auditd-plugin
+
+    Copyright (c) Microsoft Corporation
+
+    All rights reserved.
+
+    MIT License
+
+    Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the ""Software""), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+#include "IO.h"
+#include "Signals.h"
+
+#include <cassert>
+#include <system_error>
+
+extern "C" {
+#include <unistd.h>
+#include <fcntl.h>
+#include <poll.h>
+}
+
+bool IOBase::IsOpen()
+{
+    return _fd.load() >= 0;
+}
+
+bool IOBase::Open()
+{
+    throw std::runtime_error("WriterBase::Open: Operation Not Supported");
+}
+
+void IOBase::Close()
+{
+    int fd = _fd.load();
+    if (fd >= 0) {
+        _fd.store(-1);
+        close(fd);
+    }
+    _rclosed.store(true);
+    _wclosed.store(true);
+}
+
+void IOBase::CloseRead() {
+    _rclosed.store(true);
+    if (_wclosed.load()) {
+        Close();
+    }
+}
+
+void IOBase::CloseWrite() {
+    _wclosed.store(true);
+    if (_rclosed.load()) {
+        Close();
+    }
+}
+
+void IOBase::SetNonBlock(bool enable) {
+    int fd = _fd.load();
+    if (_fd < 0) {
+        return;
+    }
+
+    int flags;
+    if (-1 == (flags = fcntl(fd, F_GETFL, 0))) {
+        flags = 0;
+    }
+    if (enable) {
+        flags |= O_NONBLOCK;
+    } else {
+        flags &= ~O_NONBLOCK;
+    }
+    if (fcntl(fd, F_SETFL, flags) != 0) {
+        throw std::system_error(errno, std::system_category());
+    }
+}
+
+ssize_t IOBase::WaitReadable(long timeout) {
+    int fd = _fd.load();
+    if (_fd < 0 || _rclosed.load()) {
+        return CLOSED;
+    }
+    struct pollfd fds;
+    fds.fd = fd;
+    fds.events = POLLIN;
+    fds.revents = 0;
+
+    auto ret = poll(&fds, 1, static_cast<int>(timeout));
```

Comment:
I notice the timeout is -1, do we need a smaller timeout to make sure that poll will return and do not block other work?

Reply:
In cases where the timeout is -1, if the thread blocked on poll needs to be terminated, the thread is sent SIGQUIT and the poll will be interrupted and will return with errno == EINTR. See RunBase.cpp for the core thread/task management logic.